### PR TITLE
Negotiate improvements

### PR DESCRIPTION
**Pull request by vanackere**
Hi,

 This branch contains various changes / improvements to the negotiate package (FYI those changes were added in order to implement both Web SPNEGO and SASL GSS-SPNEGO support, using the native windows api).

Please merge if you feel those are appropriate, thanks !

### CODE REVIEW

**Comment by alexbrainman** [comment]
I will try to review this weekend.

Alex

File: sspi.go [hunk 1]
```diff
@@ -50,13 +50,21 @@ type Credentials struct {
 	expiry syscall.Filetime
 }
 
-func AcquireCredentials(pkgname string, creduse uint32, authdata *byte) (*Credentials, error) {
+func AcquireCredentials(principal string, pkgname string, creduse uint32, authdata *byte) (*Credentials, error) {
```
[the remark this answers is not in the document]
Comment (vanackere) [reply sspi.go]: An empty principal parameter is passed as a nil pointer in the underlying winapi call. I will document.

File: negotiate/negotiate.go [hunk 2]
```diff
@@ -245,6 +288,20 @@ func (c *ClientContext) MakeSignature(msg []byte, qop, seqno uint32) ([]byte, er
 	return makeSignature(c.sctxt, msg, qop, seqno)
 }
 
+// EncryptMessage uses the established client context to encrypt a message
```
[the remark this answers is not in the document]
Comment (vanackere) [reply negotiate/negotiate.go]: Done

**Comment by vanackere** [comment]
Hi, I made the requested changes and updated my branch, PTAL

File: negotiate/negotiate.go [hunk 3]
```diff
@@ -85,13 +85,13 @@ func AcquireUserCredentials(domain, username, password string) (*sspi.Credential
 		PasswordLength: plen,
 		Flags:          sspi.SEC_WINNT_AUTH_IDENTITY_UNICODE,
 	}
-	return acquireCredentials(sspi.SECPKG_CRED_OUTBOUND, &ai)
+	return acquireCredentials("", sspi.SECPKG_CRED_OUTBOUND, &ai)
 }
 
 // AcquireServerCredentials acquires server credentials that will
 // be used to authenticate client.
```
[the remark this answers is not in the document]
Comment (alexbrainman) [reply negotiate/negotiate.go]: You did not address my comment.

File: sspi.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply sspi.go]: I still do not see any new documentation that you promised.

File: negotiate/negotiate.go [hunk 4]
```diff
@@ -211,13 +211,23 @@ type ClientContext struct {
 	targetName *uint16
 }
 
-// NewClientContext creates new client context. It uses client
+// NewClientContext creates a new client context. It uses client
 // credentials cred generated by AcquireCurrentUserCredentials or
-// AcquireUserCredentials and SPN to start client Negotiate
+// AcquireUserCredentials and SPN to start a client Negotiate
 // negotiation sequence. targetName is the service principal name
 // (SPN) or the security context of the destination server.
```
[the remark this answers is not in the document]
Comment (alexbrainman) [reply negotiate/negotiate.go]: You did not address my comment.

File: sspi.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vanackere) [reply sspi.go]: I forgot this part, sorry. I'll fix that.

File: negotiate/negotiate.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vanackere) [reply negotiate/negotiate.go]: The flags parameter is documented in AcquireUserCredentialsWithFlags, did you mean something else ?

File: negotiate/negotiate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vanackere) [reply negotiate/negotiate.go]: I'm sorry, I'm afraid I did not understand what you meant here... Do you suggest to add a new parameter "principalName" to AcquireUserCredentials ?

File: negotiate/negotiate.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply negotiate/negotiate.go]: My fault, I have no idea why I asked you to document "flag" parameter here. Forget my comment.

File: negotiate/negotiate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply negotiate/negotiate.go]: You have added "principalName" parameter to AcquireServerCredentials function. You should document new parameter  - explain what it does.

File: sspi.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexbrainman) [reply sspi.go]: Please, do.

**Comment by vanackere** [comment]
Branch updated and rebased, please let me know if I still missed anything, thanks !

**Comment by alexbrainman** [comment]
LGTM now. Thank you.

Alex

**Comment by alexbrainman** [comment]
Merged, so closing this now.

Alex

**Pull request merged**